libs/langgraph/langgraph/pregel/_algo.py

```diff
@@ -1106,3 +1111,56 @@ def __call__(self) -> int:
                 if self._counter is None:
                     self._counter = itertools.count(0).__next__
         return self._counter()
+
+
+def sanitize_untracked_values_in_send(
+    packet: Send, channels: Mapping[str, BaseChannel]
+) -> Send:
+    """Replace any UntrackedValue contents in Send.arg with RUNTIME_PLACEHOLDER for checkpointing.
+
+    Send is not typed and arg may be a nested dict."""
+
+    if not isinstance(packet.arg, dict):
+        # Command
+        return packet
+
+    def replace(obj: dict[str, Any]) -> dict[str, Any]:
+        for k, v in obj.items():
+            if isinstance(v, dict):
+                # arg can be nested dicts
+                v = replace(v)
+            if isinstance(channels.get(k), UntrackedValue):
+                obj[k] = RUNTIME_PLACEHOLDER
+        return obj
+
+    sanitized_arg = replace(packet.arg)
+    return Send(node=packet.node, arg=sanitized_arg)
+
+
+def rehydrate_untracked_values_in_send(
+    packet: Send, channels: Mapping[str, BaseChannel]
+) -> Send:
+    """Replace RUNTIME_PLACEHOLDERs in Send.arg with actual untracked values from UntrackedValue channels."""
+
+    if not isinstance(packet.arg, dict):
+        # Command
+        return packet
+
+    # deepcopy to avoid mutating the original packet, as it is later persisted in checkpoints
```

Comment:
Deep copying is CPU and mem intensive, and we'd only everf have an untracked value at the top level of the dict, meaning we'd only need to do a shallow one. Also, we should probably avoid calling this function of no untracked values are in teh channels map.